### ISSUE TITLE
tf.GradientTape().tape() is returning None on the generator loss

### ISSUE THREAD

**Issue by sayakpaul** [tensorflow/tensorflow#31919]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: I am using Colab
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: Not applicable
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: `2.0.0-beta1`
- **Python version**: 3.7
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**: [This Colab notebook](https://colab.research.google.com/drive/1iIGWaktKZfDmxpjl4XARsBinTHcgd2Xf)

### Describe the problem

I am trying to reproduce [this simplistic GANs example](https://medium.com/@devnag/generative-adversarial-networks-gans-in-50-lines-of-code-pytorch-e81b79659e3f) using TensorFlow 2.0. Everything is working as _expected except for the generator network_. Here's one forward and backward pass of the generator network:

```python
with tf.GradientTape() as tape:
  g_fake_data = G(gen_input)
  dg_fake_decision = D(get_moments(g_fake_data.numpy().T).reshape((1,4)))
  g_error = criterion(dg_fake_decision, np.ones((1,1)))
tape.gradient(g_error, G.trainable_weights)
```
The gradients are all coming as `None`. I have tried to do `tape.watch(G.trainable_weights)` as well but it does not help. My suspect is that the `GradientTape` context manager is unable to keep track of the appropriate dependencies to compute the gradients. I have tried to move the `dg_fake_decision = D(get_moments(g_fake_data.numpy().T).reshape((1,4)))` step to `with tape.stop_recording()` but it still does not help. 

Any suggestions on this would be very helpful. I hope this would be helpful for the community specifically for them who are willing to dig deep with automatic differentiation.

**Comment by WindQAQ** [comment]
Hello, maybe official tutorial could help you.
https://www.tensorflow.org/beta/tutorials/generative/dcgan

**Comment by sayakpaul** [comment]
Hi. I am aware of this tutorial but the topologies are different here.

**Comment by WindQAQ** [comment]
After a rough look at the colab, I find `get_moments()` is a numpy function, so you pass `g_fake_data.numpy().T` into it, which makes the graph unconnected. So I modify this function to TF version and the gradients are not longer `None`. Modified colab notebook is down below:
https://colab.research.google.com/drive/1r9HZRlVepGU6ScXPZLOjwoLa7vSr5DCJ

**Comment by sayakpaul** [comment]
Thank you very much @WindQAQ. I will look into and post my further findings here :)
